src/backend/distributed/planner/multi_logical_planner.c

```diff
@@ -1993,62 +2038,107 @@ SubqueryPushdownMultiPlanTree(Query *queryTree)
 {
 	List *targetEntryList = queryTree->targetList;
 	List *qualifierList = NIL;
-	List *qualifierColumnList = NIL;
-	List *targetListColumnList = NIL;
 	List *columnList = NIL;
 	ListCell *columnCell = NULL;
 	MultiCollect *subqueryCollectNode = CitusMakeNode(MultiCollect);
 	MultiTable *subqueryNode = NULL;
-	MultiSelect *selectNode = NULL;
 	MultiProject *projectNode = NULL;
 	MultiExtendedOp *extendedOpNode = NULL;
 	MultiNode *currentTopNode = NULL;
 	RangeTblEntry *subqueryRangeTableEntry = NULL;
-	List *subqueryEntryList = SubqueryEntryList(queryTree);
-
+	Query *queryCopy = (Query *) copyObject(queryTree);
+	AttrNumber targetNo = 1;
+	List *subqueryTargetEntryList = NIL;
+	List *columnNamesList = NIL;
+	StringInfo rteName = makeStringInfo();
+	List *knownColumnList = NIL;
 
 	/* verify we can perform distributed planning on this query */
-	ErrorIfQueryNotSupported(queryTree);
-	ErrorIfSubqueryJoin(queryTree);
+	ErrorIfQueryNotSupported(queryTree, true);
 
 	/* extract qualifiers and verify we can plan for them */
 	qualifierList = QualifierList(queryTree->jointree);
 	ValidateClauseList(qualifierList);
 
 	/*
-	 * We disregard pulled subqueries. This changes order of range table list.
-	 * We do not allow subquery joins, so we will have only one range table
-	 * entry in range table list after dropping pulled subquery. For this reason,
-	 * here we are updating columns in the most outer query for where clause
-	 * list and target list accordingly.
+	 * We would be creating a new Query and pushing down top level query's
+	 * contents down to it. Join and filter clauses in higher level query would
+	 * be transferred to lower query. Therefore after this function we would
+	 * only have a single range table entry in the top level query. We need to
+	 * create a target list entry in lower query for each column reference in
+	 * upper level query's target list, and update varno/varattno attributes
+	 * of them accordingly.
 	 */
-	Assert(list_length(subqueryEntryList) == 1);
-
-	qualifierColumnList = pull_var_clause_default((Node *) qualifierList);
-	targetListColumnList = pull_var_clause_default((Node *) targetEntryList);
-
-	columnList = list_concat(qualifierColumnList, targetListColumnList);
+	columnList = pull_var_clause_default((Node *) targetEntryList);
 	foreach(columnCell, columnList)
 	{
+		TargetEntry *newTargetEntry = makeNode(TargetEntry);
```

Comment:
done for the most part.

It would have been nice if all newly introduced code to be placed in to a single function, however, since we also change queryTree, when we reassign varno/varattno fields  in target entry list and and havingQuals, I did not want to have a function with side effects.